peer/peerdiscovery.go

```diff
@@ -0,0 +1,173 @@
+	transport *transport.Transport
+}
+
+func NewDiscoveryClient(opts DiscoveryOptions, transport *transport.Transport) *DiscoveryClient {
+	return &DiscoveryClient{
+		opts: opts,
+
+		transport: transport,
+	}
+}
+
+func (dc *DiscoveryClient) DiscoverPeers(ctx context.Context) {
+	var maxExpectedPeers int
+	if dc.opts.MaxExpectedPeers <= 0  {
+		dc.opts.Logger.Debug("pinging", zap.String("max expected peers cannot be less than 1", "setting to default - 5"))
+		maxExpectedPeers = 5
+	} else {
+		maxExpectedPeers = dc.opts.MaxExpectedPeers
+	}
+
+	go func() {
+		var pingData [16]byte
+		binary.LittleEndian.PutUint64(pingData[:8], uint64(maxExpectedPeers))
+		binary.LittleEndian.PutUint64(pingData[8:], uint64(dc.transport.Port()))
+
+		msg := wire.Msg{
+			Version: wire.MsgVersion1,
+			Type:    wire.MsgTypePing,
+			Data:    pingData[:],
+		}
+		for {
+			for _, sig := range dc.transport.Table().Peers(dc.opts.Alpha) {
+				msg.To = id.Hash(sig)
+				err := dc.transport.Send(ctx, sig, msg)
+				if err != nil {
+					dc.opts.Logger.Debug("pinging", zap.Error(err))
+					dc.transport.Table().DeletePeer(sig)
+				}
+			}
+			<-time.After(dc.opts.PingTimePeriod)
+		}
+	}()
+}
+
+func (dc *DiscoveryClient) DidReceiveMessage(from id.Signatory, msg wire.Msg) {
+	switch msg.Type {
+	case wire.MsgTypePing:
+		dc.didReceivePing(from, msg)
+	case wire.MsgTypePingAck:
+		dc.didReceivePingAck(from, msg)
+	}
+}
+
+func (dc *DiscoveryClient) didReceivePing(from id.Signatory, msg wire.Msg) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var addressBuf [1024]byte
+	addressByteSlice := addressBuf[:0]
+	expNumPeer := int(binary.LittleEndian.Uint64(msg.Data[:8]))
+	port := uint16(binary.LittleEndian.Uint64(msg.Data[8:]))
+	ipAddr, _ := dc.transport.Table().IP(from)
+	dc.transport.Table().AddPeer(
+		from,
+		wire.NewUnsignedAddress(wire.TCP, fmt.Sprintf("%v:%v", ipAddr, port), uint64(time.Now().UnixNano())),
+		)
+
+	peers := dc.transport.Table().Peers(expNumPeer)
+	for _, sig := range peers {
+		var addrAndSig [128]byte
+		addrAndSigSlice := addrAndSig[:]
+		addr, ok := dc.transport.Table().PeerAddress(sig)
+		if !ok {
+			dc.opts.Logger.Debug("sending pingAck", zap.String("peer", "does not exist in table"))
+			dc.transport.Table().DeletePeer(sig)
+		}
+		sigAndAddr := SignatoryAndAddress{Signatory: sig, Address: addr}
+		tail, _, err := sigAndAddr.Marshal(addrAndSigSlice, len(addrAndSigSlice))
+		if err != nil {
+			dc.opts.Logger.Debug("sending pingAck", zap.Error(err))
+		}
+		addressByteSlice = append(addressByteSlice, addrAndSigSlice[:len(addrAndSigSlice)-len(tail)]...)
+	}
+	response := wire.Msg{
+		Version: wire.MsgVersion1,
+		Type:    wire.MsgTypePingAck,
+		To:      id.Hash(from),
+		Data:    addressByteSlice,
+	}
+	err := dc.transport.Send(ctx, from, response)
+	if err != nil {
+		dc.opts.Logger.Debug("sending pingAck", zap.Error(err))
+	}
+}
+
+func (dc *DiscoveryClient) didReceivePingAck(from id.Signatory, msg wire.Msg) {
```

Comment:
Surge is able to unmarshal directly into a slice. For example:

```
slice := []SignatoryAndAddress{}
_, _, err := surge.Unmarshal(&slice, data, surge.MaxBytes)
```

You can also adjust the `surge.MaxBytes` to ensure that the unmarshaller will not allocate too much memory.

Reply:
This doesn't seem to work